### PR TITLE
Add Converter (DLK) limitation to blueoil 

### PR DESCRIPTION
**Pull request by oatawa1**
What this patch does to fix the issue.
- Create DLK limitation documents
- Fix comment typo in `operators.py`

## Link to any relevant issues or pull requests.
#470 

## Todo
- Supported OPS
  - [x] Verify that the required `output channe`l is put in the correct part.
  - [x] Verify the limitation of `blueoil customized ops`.
  - [x] Check again for `Dropout` and `Flatten`, they are used in model but it is not shown in op_type in `tf.py: create_new_model`  
  - [x] `Gemm` is not TF command, should we move to blueoil customized ops?
- Some datatype cannot map to tf data type
  - [x] DT_INVALID
  - [x] DT_HALF
  - [x] FLOATS, INTS
  - [x] STRINGS
  - [x] TENSOR, t
  - [x] GRAPH, g
  - [x] TENSORS
  - [x] GRAPHS

### CODE REVIEW

File: blueoil/converter/core/operators.py [hunk 1]
```diff
@@ -2562,7 +2562,7 @@ def __init__(self,
     def _check_consistency(self) -> None:
         """
         This check the following constraints:
-            1. qunatized-packed data requires depth of input must be multiple of kernel_size^2 * 32
+            1. quantized-packed data requires depth of input must be multiple of kernel_size^2 * 32
```
Comment (iizukak) [inline blueoil/converter/core/operators.py]: Please fix in other PR.

File: docs/dlk/limitation.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
+# Limitation
```
Comment (iizukak) [inline docs/dlk/limitation.md]: The title is `Supported Ops` is good.
File name also.

File: docs/dlk/limitation.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+## DLK Data Types
```
Comment (iizukak) [inline docs/dlk/limitation.md]: DLK is older name of Converter

File: docs/dlk/limitation.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+### Supported data type
```
Comment (iizukak) [inline docs/dlk/limitation.md]: Theses names comes from TensorFlow internal type name.
We assume document readers are not familiar with TF internal.

File: docs/dlk/limitation.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+ - **DT_COMPLEX128**
+ - **STRINGS**
+ - **TENSOR**
+ - **GRAPH**
+ - **t**
+ - **g**
+ - **TENSORS**
+ - **GRAPHS**
+ 
+## DLK Operators with limitations
```
Comment (iizukak) [inline docs/dlk/limitation.md]: DLK -> Converter

File: docs/dlk/limitation.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+### Base class
```
Comment (iizukak) [inline docs/dlk/limitation.md]: Base class seems not TensorFlow Op name.

File: docs/dlk/limitation.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+ - **Input**
+ - **Output**
+    - Require each output channel size <= `1024`
+ - **Constant**
+ - **Lookup**
```
Comment (iizukak) [inline docs/dlk/limitation.md]: Need new line here.

File: docs/dlk/limitation.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+ - **AveragePool** (AvgPool)
+    - Currently, support only `2D`
+
+ - **Add** (AddV2, BiasAdd)
+    
+ - **ConcatOnDepth** (ConcatV2)
+ 
+ - **Gather** (GatherV2)
+ 
+ - **BinaryChannelWiseMeanScalingQuantizer** (QTZ_binary_channel_wise_mean_scaling)
```
Comment (iizukak) [inline docs/dlk/limitation.md]: `BinaryChannelWiseMeanScalingQuantizer ` and `BinaryMeanScalingQuantizer` is our custom Op. Please write that.

File: docs/dlk/limitation.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+ - **GRAPHS**
+ 
+## DLK Operators with limitations
+### Base class
+ - **Input**
+ - **Output**
+    - Require each output channel size <= `1024`
+ - **Constant**
+ - **Lookup** 
+### Tensorflow operators with name substitution
```
Comment (iizukak) [inline docs/dlk/limitation.md]: What `substitution` means?

**Comment by oatawa1** [comment]
@iizukak Thank you for the review, I will edit as your suggestions.

File: blueoil/converter/core/operators.py (hunk 1 above)
Comment (oatawa1) [reply blueoil/converter/core/operators.py]: Move to #926

File: docs/dlk/supported_ops.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    - Supports only `channel-wise paddings`.
+- **[tf.nn.space_to_depth](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/nn/space_to_depth)**
+    - Requires `output depth = (multiple of kernel_size^2 * 32)` or `(kernel_size^2 * {8, 16})`.
+- **[tf.split](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/split)**
+    - Currently, all of output tensors must have `same` shape.
+    - For quantized tensor, requires `number of channel of each output tensor = multiple of 32`.
+
+###  Tensorflow Supported without Limitations
+- **[tf.math.add](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/math/add)**
+- **[tf.cast](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/cast)**
```
Comment (primenumber) [inline docs/dlk/supported_ops.md]: This operator is not supported by runtime.

File: docs/dlk/supported_ops.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+- **[tf.gather](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/gather)**
```
Comment (primenumber) [inline docs/dlk/supported_ops.md]: This operator is not supported by runtime but used for pass_lookup.

**Comment by oatawa1** [comment]
@primenumber Thanks. I'm going to fix it.

File: docs/dlk/supported_ops.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+- **[tf.layers.Conv2D](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/layers/Conv2D)**
+    - Support only convolution `2D`.
+    - Requires `kernel size = 1x1` or `3x3`.
+    - Requires `Input/output channel size = multiple of 32`, otherwise zero padding is used.
+    - **Do not** support transpose.
+- **[tf.nn.depth_to_space](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/nn/depth_to_space)**
+    - Requires `depth of input = multiple of kernel_size^2 * 32`.
+- **[tf.nn.fused_batch_norm](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/nn/fused_batch_norm)**
+    - Batch normalization just before quantization is folded with quantization.
+    - Currently this function is only used by threshold skipping optimization pass for recursively calculating thresholds of the skipping patterns.
```
[the remark this answers is not in the document]
Comment (primenumber) [reply docs/dlk/supported_ops.md]: `Currently this function is only used by threshold skipping optimization pass for recursively calculating thresholds of the skipping patterns` is comment for `BatchNormalization::run()`, not for its class.
True limitation is `scale` `B` `mean` `var` `epsilon` must be constants or computable from constants.

[the remark this answers is not in the document]
Comment (oatawa1) [reply docs/dlk/supported_ops.md]: Thank you. 
May I confirm my understanding, is `B` an `offset`?

[the remark this answers is not in the document]
Comment (primenumber) [reply docs/dlk/supported_ops.md]: Yes.

[the remark this answers is not in the document]
Comment (oatawa1) [reply docs/dlk/supported_ops.md]: Great, thanks. I fixed it.

File: docs/converter/supported_ops.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+- **[tf.layers.AveragePooling2D](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/layers/AveragePooling2D)**
+    - Currently, support only `2D`.
+    - **Do not** support `kernel depth = 1`.
+- **[tf.concat](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/concat)**
+    - **Do not** support concat of mixed data types (e.g., quantized values and float values).
+    - If inputs are quantized, requires `Each input channel size = multiple of 32`.
+- **[tf.layers.Conv2D](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/layers/Conv2D)**
+    - Support only convolution `2D`.
+    - **Do not** support transpose.
+    - Requires `kernel size = 1x1` or `3x3` or `5x5`.
```
Comment (iizukak) [inline docs/converter/supported_ops.md]: Our accelerator dose not support 5x5. Please write that.

File: docs/converter/supported_ops.md [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    - Requires `Input channel size = multiple of 32`, otherwise zero padding is used.
+    - If output is quantized by later operations, `Output channel size = multiple of 32`, otherwise output channel size is free from limitation (but performance will be worse).
+- **[tf.nn.depth_to_space](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/nn/depth_to_space)**
```
Comment (iizukak) [inline docs/converter/supported_ops.md]: Sorry, what is the kernel for s2d?

File: docs/converter/supported_ops.md [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+- **QTZ_binary_channel_wise_mean_scaling**
+- **QTZ_binary_mean_scaling**
+- **QTZ_linear_mid_tread_half**
+
+### Tensorflow Supported with Limitations
+- **[tf.layers.AveragePooling2D](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/layers/AveragePooling2D)**
+    - Currently, support only `2D`.
+    - **Do not** support `kernel depth = 1`.
+- **[tf.concat](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/concat)**
+    - **Do not** support concat of mixed data types (e.g., quantized values and float values).
```
[the remark this answers is not in the document]
Comment (oatawa1) [reply docs/converter/supported_ops.md]: Thank you for raise this up.
I think not only the channel size but all dimensions except axis.

According to tf document [here](https://www.tensorflow.org/api_docs/python/tf/concat), 
`The number of dimensions of the input tensors must match, and all dimensions except axis must be equal.`

However, this limitation is 
Currently, the limitation only from blueoil side is listed.
Should we add the general limitation from tensorflow as well?

File: docs/converter/supported_ops.md (hunk 14 above)
Comment (oatawa1) [reply docs/converter/supported_ops.md]: My understanding is the kernel size in block_size for either s2d or d2s.

Comment (iizukak) [reply docs/converter/supported_ops.md]: Then how about to write `block_size ` instead of `kernel`.

File: docs/converter/supported_ops.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (iizukak) [reply docs/converter/supported_ops.md]: Thanks.

File: docs/converter/supported_ops.md (hunk 14 above)
Comment (oatawa1) [reply docs/converter/supported_ops.md]: Fixed

File: docs/converter/supported_ops.md [hunk 16]
```diff
@@ -0,0 +1,93 @@
+# Supported Ops
+## Converter Ops with Limitations
+### Base Limitations
+- **Output**
+    - Requires each output channel size <= `1024`.
+
+### Blueoil Customized Ops
+- **QTZ_binary_channel_wise_mean_scaling**
```
[the remark this answers is not in the document]
Comment (oatawa1) [reply docs/converter/supported_ops.md]: Link added

**Comment by yd8534976** [comment]
Thanks! It seems `ResizeNearestNeighbor` is missed out.

**Comment by oatawa1** [comment]
@yd8534976 Thanks, my mistake I will add it.

**Comment by iizukak** [comment]
@oatawa1 Of couse.
